server/src/test/java/org/elasticsearch/cluster/coordination/AtomicRegisterCoordinatorTests.java

```diff
@@ -0,0 +1,693 @@
+            final var currentRegisterPersistentState = register.readClusterState(key);
+            if (currentRegisterPersistentState == null) {
+                return false;
+            }
+            return currentRegisterPersistentState.term() > localAcceptedTerm
+                || (currentRegisterPersistentState.term() == localAcceptedTerm
+                    && currentRegisterPersistentState.version() > localAcceptedVersion);
+        }
+
+        @Override
+        public boolean isPublishQuorum(
+            CoordinationState.VoteCollection voteCollection,
+            CoordinationMetadata.VotingConfiguration lastCommittedConfiguration,
+            CoordinationMetadata.VotingConfiguration latestPublishedConfiguration
+        ) {
+            assert latestPublishedConfiguration.getNodeIds().size() == 1;
+
+            return voteCollection.isQuorum(latestPublishedConfiguration);
+        }
+
+        @Override
+        public boolean shouldJoinLeaderInTerm(long currentTerm, long targetTerm, Optional<Join> lastJoin) {
+            // Ensure that when we become candidate, and we've read the cluster state from a register we update our info
+            return currentTerm < targetTerm || currentTerm == targetTerm && lastJoin.isEmpty();
+        }
+
+        @Override
+        public boolean isValidStartJoinRequest(StartJoinRequest startJoinRequest, long currentTerm) {
+            // Bootstrap - we have the same term as the leader
+            // During a race where we lose the election using the same term, we should join
+            return startJoinRequest.getTerm() >= currentTerm;
+        }
+
+        @Override
+        public void onNewElection(DiscoveryNode sourceNode, long proposedTerm) {
+            if (register.claimTerm(key, sourceNode, proposedTerm) == false) {
+                throw new CoordinationStateRejectedException("Term " + proposedTerm + " already claimed by another node");
+            }
+        }
+    }
+
+    record PersistentClusterState(long term, long version, Metadata state) {
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PersistentClusterState that = (PersistentClusterState) o;
+            return term == that.term
+                && version == that.version
+                && Objects.equals(state.coordinationMetadata(), that.state().coordinationMetadata());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(term, version, state.coordinationMetadata());
+        }
+    }
+
+    record TermOwner(DiscoveryNode node, long term) {
+        static TermOwner EMPTY = new TermOwner(null, 0);
+    }
+
+    static class KeyedAtomicRegister {
+        private final Map<String, PersistentClusterState> clusterStates = new HashMap<>();
+        private final Map<String, HeartBeat> heartBeats = new HashMap<>();
```

Comment:
Likewise with heartbeats, I don't think we need (want?) them to be register-based. They can also be per-term things, with each node watching the liveness blob for its current term.